lib/reversetunnel/agentpool.go

```diff
@@ -356,12 +366,29 @@ func (l ServerHandlerToListener) Close() error {
 }
 
 func (l ServerHandlerToListener) Addr() net.Addr {
-	return reverseTunnelAddr{}
+	return reverseTunnelAddr(l.tunnelAddr)
+}
+
+type connCloser struct {
+	net.Conn
+	closeOnce *sync.Once
+	closed    chan struct{}
 }
 
+func newConnCloser(c net.Conn) connCloser {
+	return connCloser{Conn: c, closeOnce: new(sync.Once), closed: make(chan struct{})}
+}
+
+func (c connCloser) Close() error {
+	c.closeOnce.Do(func() { close(c.closed) })
+	return c.Conn.Close()
+}
+
+func (c connCloser) wait() { <-c.closed }
+
 // reverseTunnelAddr is a net.Addr implementation for a listener based on a
 // reverse tunnel.
-type reverseTunnelAddr struct{}
+type reverseTunnelAddr string
 
-func (reverseTunnelAddr) Network() string { return "ssh" }
-func (reverseTunnelAddr) String() string  { return "ssh://kube.reversetunnel.teleport.cluster.local" }
+func (reverseTunnelAddr) Network() string  { return "ssh-reversetunnel" }
```

Comment:
Not sure yet how exactly this is used in the reverse tunnel subsystem (if at all), but just making sure this won't cause backwards-incompatibility b/w old and new versions of components.

Reply:
this code was added in https://github.com/gravitational/teleport/pull/4611 and is not used by any components other than `kubernetes_service`.
until we release 5.0, there are no backwards-compatibility concerns for `kubernetes_service` since it's new.